service-proxy.js

```diff
@@ -321,11 +328,100 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
 function refreshServicePeer(serviceName, hostPort) {
     var self = this;
 
+    // Create a peer for the worker.
+    // This is necessary for populating the worker pool regardless of whether
+    // we connect.
+    // TODO This belies an underlying assumption that the peer pool includes
+    // exactly and only the worker pool, that we would eventually reap peers
+    // that have not advertised recently.
     var peer = self.getServicePeer(serviceName, hostPort);
-    peer.connectTo();
 
+    // Reset the expiration time for this service peer
     var time = self.channel.timers.now();
     self.exitServices[serviceName] = time;
+
+    if (self.partialAffinityEnabled) {
+        return self.refreshServicePeerPartially(serviceName, hostPort);
+    }
+
+    // The old way: fully connect every egress to all affine peers.
+    peer.connectTo();
+};
+
+ServiceDispatchHandler.prototype.refreshServicePeerPartially =
+function refreshServicePeerPartially(serviceName, hostPort) {
+    var self = this;
+
+    // Obtain and sort the affine worker and relay lists.
+    var relays = Object.keys(self.egressNodes.exitsFor(serviceName));
+    relays.sort();
+    var serviceChannel = self.getOrCreateServiceChannel(serviceName);
+    var workers = serviceChannel.peers.keys();
+    workers.sort();
+
+    // Find our position within the affine relay set so we can project that
+    // position into the affine worker set.
+    var relayIndex = sortedIndexOf(relays, self.channel.hostPort);
+    // istanbul ignore if
+    if (relayIndex < 0) {
+        // This should only occur if an advertisement loses the race with a
+        // relay ring membership change.
+        self.logger.warn('Relay could not find itself in the affinity set for service', self.extendLogInfo({
+            serviceName: serviceName,
+            relayHostPort: self.channel.hostPort,
+            workerHostPort: hostPort,
+            relays: relays,
+            workers: workers
+        }));
+        return;
+    }
+
+    // Compute the range of workers that this relay should be connected to.
+    var ratio = workers.length / relays.length;
+    var start = Math.floor(relayIndex * ratio);
+    var length = Math.ceil(
+        Math.min(
+            workers.length,
+            Math.max(
+                self.minPeersPerRelay,
+                self.minPeersPerWorker * ratio
+            )
+        )
+    );
+    var stop = (start + length) % workers.length;
+
+    self.logger.info('Refreshing service peer affinity', self.extendLogInfo({
```

Comment:
This log is O(n) advertise. 

Once we've proven the feature with confidence we don't want this.

We may want a log for "new peers" or "removing peers" though.

Hyperbahn should be silent under steady state. We can keep this here for now.